drivers/espi/espi_taf_npcx.c

```diff
@@ -0,0 +1,470 @@
+/*
+ * Copyright (c) 2023 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_espi_taf
+
+#include <soc.h>
+#include <zephyr/drivers/espi.h>
+#include <zephyr/drivers/espi_saf.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(espi_taf, CONFIG_ESPI_LOG_LEVEL);
+
+static const struct device *const spi_dev = DEVICE_DT_GET(DT_ALIAS(taf_flash));
+
+struct espi_taf_npcx_config {
+	uintptr_t base;
+	uintptr_t mapped_addr;
+	uintptr_t rx_plsz;
+	enum NPCX_ESPI_TAF_ERASE_BLOCK_SIZE erase_sz;
+	enum NPCX_ESPI_TAF_MAX_READ_REQ max_rd_sz;
+};
+
+struct espi_taf_npcx_data {
+	sys_slist_t callbacks;
+};
+
+#define HAL_INSTANCE(dev)						\
+	((struct espi_reg *)((const struct espi_taf_npcx_config *)	\
+	(dev)->config)->base)
+
+#define DT_INST_ERASE_SZ_PROP_OR(inst)					\
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, erase_sz),		\
+		    (_CONCAT(NPCX_ESPI_TAF_ERASE_BLOCK_,		\
+		     DT_INST_STRING_TOKEN(inst, erase_sz))),		\
+		    ((NPCX_ESPI_TAF_ERASE_BLOCK_SIZE_4KB)))
+
+#define DT_INST_MAX_READ_SZ_PROP_OR(inst)				\
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, max_read_sz),		\
+		    (_CONCAT(NPCX_ESPI_TAF_MAX_READ_REQ_,		\
+		     DT_INST_STRING_TOKEN(inst, max_read_sz))),		\
+		    ((NPCX_ESPI_TAF_MAX_READ_REQ_64B)))
+
+/* Check access region of read request is protected or not */
+bool espi_taf_check_read_protect(const struct device *dev, uint32_t addr, uint32_t len,
+				 uint8_t tag)
+{
+	struct espi_reg *const inst = HAL_INSTANCE(dev);
+	uint32_t flash_addr = addr;
+	uint8_t i;
+	uint16_t override_rd;
+	uint32_t base, high;
+	bool rdpr;
+
+	flash_addr +=
+		GET_FIELD(inst->FLASHBASE, NPCX_FLASHBASE_FLBASE_ADDR)
+		<< GET_FIELD_POS(NPCX_FLASHBASE_FLBASE_ADDR);
+
+	for (i = 0; i < CONFIG_ESPI_TAF_PR_NUM; i++) {
+		base = GET_FIELD(inst->FLASH_PRTR_BADDR[i],
+				 NPCX_FLASH_PRTR_BADDR)
+				 << GET_FIELD_POS(NPCX_FLASH_PRTR_BADDR);
+		high = GET_FIELD(inst->FLASH_PRTR_HADDR[i],
+				 NPCX_FLASH_PRTR_HADDR)
+				 << GET_FIELD_POS(NPCX_FLASH_PRTR_HADDR);
```

Comment:
may be worth doing a couple access macros for these, just in this file so they don't spill over the generic headers, maybe something like

```c
#define FLBASE_ADDR ( \
        GET_FIELD(inst->FLASHBASE, NPCX_FLASHBASE_FLBASE_ADDR) \
        << GET_FIELD_POS(NPCX_FLASHBASE_FLBASE_ADDR))

#define PRTR_BADDR(i) ( \
        GET_FIELD(inst->FLASH_PRTR_BADDR[i], NPCX_FLASH_PRTR_BADDR) \
        << GET_FIELD_POS(NPCX_FLASH_PRTR_BADDR))

#define PRTR_HADDR(i) ( \
        GET_FIELD(inst->FLASH_PRTR_HADDR[i], NPCX_FLASH_PRTR_HADDR) \
        << GET_FIELD_POS(NPCX_FLASH_PRTR_HADDR))
```

Reply:
Done.